spring-cloud-kubernetes-discovery/src/main/java/org/springframework/cloud/kubernetes/discovery/KubernetesDiscoveryClient.java

```diff
@@ -14,132 +14,161 @@
-		return client.services().list()
-			.getItems()
-			.stream().map(s -> s.getMetadata().getName())
-			.collect(Collectors.toList());
-	}
+    private static final Log log = LogFactory.getLog(KubernetesDiscoveryClient.class);
+    private static final String HOSTNAME = "HOSTNAME";
+
+    private KubernetesClient client;
+    private final KubernetesDiscoveryProperties properties;
+    private final SpelExpressionParser parser = new SpelExpressionParser();
+    private final SimpleEvaluationContext evalCtxt = SimpleEvaluationContext
+                                                        .forReadOnlyDataBinding()
+                                                        .withInstanceMethods()
+                                                        .build();
+
+    public KubernetesDiscoveryClient(KubernetesClient client,
+            KubernetesDiscoveryProperties kubernetesDiscoveryProperties) {
+
+        this.client = client;
+        this.properties = kubernetesDiscoveryProperties;
+    }
+
+    public KubernetesClient getClient() {
+        return client;
+    }
+
+    public void setClient(KubernetesClient client) {
+        this.client = client;
+    }
+
+    @Override
+    public String description() {
+        return "Kubernetes Discovery Client";
+    }
+
+    public ServiceInstance getLocalServiceInstance() {
+        String serviceName = properties.getServiceName();
+        String podName = System.getenv(HOSTNAME);
+        ServiceInstance defaultInstance = new DefaultServiceInstance(serviceName,
+                "localhost",
+                8080,
+                false);
+
+        Endpoints endpoints = client.endpoints().withName(serviceName).get();
+        Optional<Service> service = Optional.ofNullable(client.services().withName(serviceName).get());
+        final Map<String, String> labels;
+        if (service.isPresent()) {
+            labels = service.get().getMetadata().getLabels();
+        } else {
+            labels = null;
+        }
+        if (Utils.isNullOrEmpty(podName) || endpoints == null) {
+            return defaultInstance;
+        }
+        try {
+            List<EndpointSubset> subsets = endpoints.getSubsets();
+
+            if (subsets != null) {
+                for (EndpointSubset s : subsets) {
+                    List<EndpointAddress> addresses = s.getAddresses();
+                    for (EndpointAddress a : addresses) {
+                        return new KubernetesServiceInstance(serviceName,
+                                a,
+                                s.getPorts().stream().findFirst().orElseThrow(IllegalStateException::new),
+                                labels,
+                                false);
+                    }
+                }
+            }
+            return defaultInstance;
+
+        } catch (Throwable t) {
+            return defaultInstance;
+        }
+    }
+
+    @Override
+    public List<ServiceInstance> getInstances(String serviceId) {
```

Comment:
Hi guys, having a look at this method and getLocalServiceInstance(), I think the logic is pretty similar. Maybe it would be nice reusing the common logic. 

If you want, I could do that in a follow up PR.

Reply:
@paolocarta I was thinking that we can remove it completely. I cannot understand why we need that besides some kind of edge case in minikube.